### PR TITLE
feat(airbyte-cdk): add global_state => per_partition transformation

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolve https://github.com/airbytehq/airbyte-internal-issues/issues/9576

## How

 add global_state => per partition transformation

## Review guide

1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py`
2. `airbyte-cdk/python/unit_tests/sources/declarative/incremental/test_per_partition_cursor.py`

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by artem1205** [comment]
Tested on source-klaviyo:
run `read` with state:
```json
"stream": {
      "stream_state": {
        "updated": "2120-10-10T00:00:00+00:00"
      },
      "stream_descriptor": { "name": "lists_detailed" }
    }
```
got transformed output per partition:

```json
{
        "states": [
          {
            "partition": { "id": "R2p3ry", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "R4ZhCr", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "RPfQMj", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "RgS4w6", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "RnsiHB", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "RwKPyg", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "S7aBY2", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "S8nmQ9", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "SBYgiK", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "SYEFFb", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "Seq8wh", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "SmDD4y", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "TDGJsj", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "TWcKFn", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "TaSce6", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "TjbH4K", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "TpNXq9", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "UeGLUr", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "UzdNhZ", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "VDZnQt", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "VJCDbR", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "VmvmBq", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "WBxsQE", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "WJLXnV", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "X7UeXn", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "XGj3p8", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "XUbNgM", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "XpP2a5", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          },
          {
            "partition": { "id": "Ya5ziX", "parent_slice": {} },
            "cursor": { "updated": "2120-10-10T00:00:00+00:00" }
          }
        ]
      }
```

**Review by tolik0 — APPROVED** [review]

File: airbyte-cdk/python/unit_tests/sources/declarative/incremental/test_per_partition_cursor.py [hunk 1]
```diff
@@ -519,10 +517,19 @@ def test_get_stream_state_includes_parent_state(mocked_cursor_factory, mocked_pa
...
+        MockedCursorBuilder().with_stream_state({CURSOR_STATE_KEY: "first slice cursor value"}).build(),
+        MockedCursorBuilder().with_stream_state({CURSOR_STATE_KEY: "second slice cursor value"}).build(),
+    ]
     cursor = PerPartitionCursor(mocked_cursor_factory, mocked_partition_router)
 
-    with pytest.raises(AirbyteTracedException) as exception:
-        cursor.set_initial_state({"invalid_state": 1})
-
-    assert exception.value.failure_type == FailureType.config_error
+    cursor.set_initial_state({"global_state_format_key": "global_state_format_value"})
```
Comment (maxi297) [inline airbyte-cdk/python/unit_tests/sources/declarative/incremental/test_per_partition_cursor.py:531]: I'm not sure I understand this test. It feels like if we were to change `cursor.set_initial_state({"global_state_format_key": "global_state_format_value"})` to `cursor.set_initial_state({"states": []})`, the test would still pass which makes me think it isn't testing global state. Should we `assert mocked_cursor_factory.create.calls == [calls(<global state>)] * <number of times we expect it to be called>`?

**Review by maxi297 — APPROVED** [review]
I would like the test to be improved. The rest is nit. I'll approve even then because it is a pretty solid PR

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 2]
```diff
@@ -56,7 +55,10 @@ def stream_slices(self) -> Iterable[StreamSlice]:
         slices = self._partition_router.stream_slices()
         for partition in slices:
             cursor = self._cursor_per_partition.get(self._to_partition_key(partition.partition))
-            if not cursor:
+            if not cursor and self._state_to_migrate_from:
+                cursor = self._create_cursor(self._state_to_migrate_from)
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:59]: > How do we differentiate from single value cursor like {"updated": "2120-10-10T00:00:00+00:00"} and just bad cursors like {"updated": "potato"}?

This is a problem to which I have no answer.

Maybe we can/should somehow set the cursor_expected_format for the cursor class, but it sounds like an unnecessary complication

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 3]
```diff
@@ -56,7 +55,10 @@ def stream_slices(self) -> Iterable[StreamSlice]:
         slices = self._partition_router.stream_slices()
         for partition in slices:
             cursor = self._cursor_per_partition.get(self._to_partition_key(partition.partition))
-            if not cursor:
+            if not cursor and self._state_to_migrate_from:
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:58]: agreed, ref

File: airbyte-cdk/python/unit_tests/sources/declarative/incremental/test_per_partition_cursor.py (hunk 1 above)
Comment (artem1205) [reply airbyte-cdk/python/unit_tests/sources/declarative/incremental/test_per_partition_cursor.py:531]: refactored, added test for call_args

Comment (maxi297) [reply airbyte-cdk/python/unit_tests/sources/declarative/incremental/test_per_partition_cursor.py:531]: Has this change been pushed? I don't see it

**Review by maxi297 — APPROVED** [review]
LGTM! Thanks for tackling the comments and being diligent about this change

**Review by girarda — APPROVED** [review]
this is great thanks @artem1205 !

@lazebnyi can you use this to avoid the breaking change in the linkedin ads migration?

**Comment by lazebnyi** [comment]
I tested it locally with https://github.com/airbytehq/airbyte/pull/44370, and everything looks good.

image

**Review by lazebnyi — APPROVED** [review]

**Pull request merged**